File: apps/android/app/src/main/java/dev/flextranslate/foundation/ModelDownloadManager.kt
package dev.flextranslate.foundation

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import androidx.compose.runtime.State
import androidx.compose.runtime.mutableStateOf
import java.io.File
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Реальная in-app загрузка моделей. Тащит файлы пака с их URL в тот же внутренний корень
 * `filesDir/models/<modelId>/`, который читают [AsrModelStore] / [MtModelStore] — так завершённая
 * загрузка сразу видна рантайму, без bundling в APK и без `adb push`.
 *
 * На каждый пак отдаёт наблюдаемое [DownloadState] (Compose [State]), которое рисует экран моделей:
 * idle / downloading (с суммарным %/байтами) / done / failed / cancelled. Загрузка идёт в фоновом
 * потоке; вся тяжёлая механика сеть/диск/проверка — в [ModelDownloadEngine].
 *
 * Только онлайн: [isOnline] честно отказывает старту, если сети нет. Идемпотентно: уже проверенный
 * пак сразу даёт [DownloadState.Done].
 */
class ModelDownloadManager(
    context: Context,
    private val resolveModelDir: (modelId: String) -> File,
    private val engine: ModelDownloadEngine = ModelDownloadEngine(),
) {

    private val appContext = context.applicationContext
    private val connectivity =
        appContext.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager

    /** Наблюдаемое состояние на пак, создаётся лениво — UI может дёрнуть `getState(id)` для любого пака. */
    private val states = ConcurrentHashMap<String, androidx.compose.runtime.MutableState<DownloadState>>()
    private val cancelFlags = ConcurrentHashMap<String, AtomicBoolean>()

    /** Куда кладётся загрузка и откуда удаляет delete — разрешённый каталог модели. */
    sealed interface DownloadState {
        data object Idle : DownloadState
        data class Downloading(val bytesDone: Long, val bytesTotal: Long, val currentFile: String?) : DownloadState {
            val fraction: Float
                get() = if (bytesTotal <= 0L) 0f else (bytesDone.toFloat() / bytesTotal.toFloat()).coerceIn(0f, 1f)
        }
        data object Done : DownloadState
        data object Cancelled : DownloadState
        data class Failed(val message: String) : DownloadState
    }

    /** True, когда есть рабочее, прошедшее валидацию сетевое соединение. */
    fun isOnline(): Boolean {
        val cm = connectivity ?: return false
        val network = cm.activeNetwork ?: return false
        val caps = cm.getNetworkCapabilities(network) ?: return false
        return caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
            caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
    }

    /** Наблюдаемое состояние для [modelId]; по умолчанию [DownloadState.Idle]. */
    fun state(modelId: String): State<DownloadState> = stateHolder(modelId)

    private fun stateHolder(modelId: String) =
        states.getOrPut(modelId) { mutableStateOf(DownloadState.Idle) }

    fun isDownloading(modelId: String): Boolean =
        stateHolder(modelId).value is DownloadState.Downloading

    /**
     * Запустить (или возобновить) загрузку [modelId]. Ничего не делает, если уже качается. Честно
     * отказывает, если нет сети или для id нет спеки загрузки. Работает в фоновом потоке.
     */
    fun start(modelId: String) {
        val spec = ModelDownloadSpecs.forModelId(modelId) ?: run {
            stateHolder(modelId).value = DownloadState.Failed("Нет источника загрузки для $modelId")
            return
        }
        val holder = stateHolder(modelId)
        if (holder.value is DownloadState.Downloading) return
        if (!isOnline()) {
            holder.value = DownloadState.Failed("Нет сети — загрузка доступна только онлайн")
            return
        }
        val cancelFlag = AtomicBoolean(false)
        cancelFlags[modelId] = cancelFlag
        holder.value = DownloadState.Downloading(0L, spec.totalBytes, null)

        Thread({
            val result = engine.download(
                spec = spec,
                modelDir = resolveModelDir(modelId),
                cancelled = cancelFlag,
                onProgress = { progress ->
                    // Не затираем терминальное состояние запоздавшим тиком прогресса после отмены.
                    if (holder.value is DownloadState.Downloading) {
                        holder.value = DownloadState.Downloading(
                            bytesDone = progress.bytesDone,
                            bytesTotal = progress.bytesTotal,
                            currentFile = progress.currentFile,
                        )
                    }
                },
            )
            holder.value = when (result) {
                ModelDownloadEngine.Result.Success -> DownloadState.Done
                ModelDownloadEngine.Result.Cancelled -> DownloadState.Cancelled
                is ModelDownloadEngine.Result.Failure -> DownloadState.Failed(result.message)
            }
            cancelFlags.remove(modelId)
        }, "flex-model-dl-$modelId").start()
    }

    /** Кооперативно отменить идущую загрузку. Частичный `.part` оставляем — пригодится для возобновления. */
    fun cancel(modelId: String) {
        cancelFlags[modelId]?.set(true)
    }

    /** Вернуть терминальный (done/failed/cancelled) пак в idle, чтобы UI снова предложил загрузку. */
    fun reset(modelId: String) {
        val holder = stateHolder(modelId)
        if (holder.value !is DownloadState.Downloading) holder.value = DownloadState.Idle
    }

    /**
     * Удалить файлы установленного пака (и оставшиеся `.part`), освободив место. Отказывает, пока идёт
     * загрузка. Возвращает true, если каталога после этого больше нет.
     */
    fun delete(modelId: String): Boolean {
        if (isDownloading(modelId)) return false
        val dir = resolveModelDir(modelId)
        val removed = dir.deleteRecursively()
        stateHolder(modelId).value = DownloadState.Idle
        return removed && !dir.exists()
    }
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/ModelDownloadManager.swift
import Foundation
import Network

/// Менеджер загрузки моделей прямо в приложении. Тянет файлы пака с их исходных URL в тот
/// же корень Application Support/models/<modelId>/, откуда читают AsrModelStore / MtModelStore,
/// так что завершённая загрузка сразу видна рантайму.
///
/// На каждый пак отдаёт наблюдаемый DownloadState, который рисует экран Models:
/// idle / downloading (с суммарными %/байтами) / done / failed / cancelled.
/// Загрузка идёт в фоновом потоке; тяжёлая механика сеть/диск/проверка — в ModelDownloadEngine.
///
/// Только онлайн: isOnline() честно отказывает в старте, если сети нет.
/// Идемпотентно: уже проверенный на месте пак сразу переходит в .done.
///
/// Калька с Android ModelDownloadManager.
@MainActor
final class ModelDownloadManager: ObservableObject {

    static let shared = ModelDownloadManager()

    private let engine: ModelDownloadEngine
    private let monitor = NWPathMonitor()
    private var currentPath: NWPath?

    @Published private(set) var states: [String: DownloadState] = [:]
    private var cancelFlags: [String: () -> Void] = [:]

    enum DownloadState: Equatable {
        case idle
        case downloading(bytesDone: Int64, bytesTotal: Int64, currentFile: String?)
        case done
        case cancelled
        case failed(String)

        var fraction: Float {
            if case .downloading(let done, let total, _) = self, total > 0 {
                return Float(done) / Float(total)
            }
            return 0
        }

        static func == (lhs: DownloadState, rhs: DownloadState) -> Bool {
            switch (lhs, rhs) {
            case (.idle, .idle): return true
            case (.done, .done): return true
            case (.cancelled, .cancelled): return true
            case (.downloading(let a, let b, let c), .downloading(let x, let y, let z)):
                return a == x && b == y && c == z
            case (.failed(let a), .failed(let b)): return a == b
            default: return false
            }
        }
    }

    init(engine: ModelDownloadEngine = ModelDownloadEngine()) {
        self.engine = engine
        monitor.pathUpdateHandler = { [weak self] path in
            Task { @MainActor in
                self?.currentPath = path
            }
        }
        monitor.start(queue: DispatchQueue.global(qos: .utility))
    }

    /// true, когда есть рабочее подтверждённое сетевое соединение.
    func isOnline() -> Bool {
        guard let path = currentPath else {
            // Монитор ещё не сработал — на всякий случай считаем, что офлайн.
            return false
        }
        return path.status == .satisfied
    }

    func state(for modelId: String) -> DownloadState {
        states[modelId] ?? .idle
    }

    func isDownloading(_ modelId: String) -> Bool {
        if case .downloading = states[modelId] { return true }
        return false
    }

    /// Запускает (или возобновляет) загрузку modelId. Ничего не делает, если уже качается.
    /// Честно отказывает в офлайне или когда для id модели нет спеки загрузки.
    func start(modelId: String) {
        guard let spec = ModelDownloadSpecs.forModelId(modelId) else {
            states[modelId] = .failed("No download source configured for \(modelId)")
            return
        }
        if isDownloading(modelId) { return }
        guard isOnline() else {
            states[modelId] = .failed("No network — download is online only")
            return
        }

        states[modelId] = .downloading(bytesDone: 0, bytesTotal: spec.totalBytes, currentFile: nil)

        var isCancelled = false
        cancelFlags[modelId] = { isCancelled = true }

        let modelDir = resolveModelDir(modelId: modelId)
        let engineRef = engine

        Task.detached(priority: .userInitiated) { [weak self] in
            let result = engineRef.download(
                spec: spec,
                modelDir: modelDir,
                cancelled: { isCancelled }
            ) { progress in
                Task { @MainActor in
                    guard let self else { return }
                    // Не затираем терминальное состояние запоздавшим тиком прогресса после отмены.
                    if case .downloading = self.states[modelId] {
                        self.states[modelId] = .downloading(
                            bytesDone: progress.bytesDone,
                            bytesTotal: progress.bytesTotal,
                            currentFile: progress.currentFile
                        )
                    }
                }
            }
            await MainActor.run { [weak self] in
                guard let self else { return }
                switch result {
                case .success:
                    self.states[modelId] = .done
                case .cancelled:
                    self.states[modelId] = .cancelled
                case .failure(let message):
                    self.states[modelId] = .failed(message)
                }
                self.cancelFlags.removeValue(forKey: modelId)
            }
        }
    }

    /// Кооперативно отменяет идущую загрузку. Недокачанный .part оставляем для возобновления.
    func cancel(modelId: String) {
        cancelFlags[modelId]?()
    }

    /// Возвращает терминальный пак (done/failed/cancelled) обратно в idle.
    func reset(modelId: String) {
        if !isDownloading(modelId) {
            states[modelId] = .idle
        }
    }

    /// Удаляет файлы установленного пака (и оставшиеся .part), освобождая место.
    /// Отказывает, пока идёт загрузка. Возвращает true, когда каталог больше не существует.
    @discardableResult
    func delete(modelId: String) -> Bool {
        if isDownloading(modelId) { return false }
        let dir = resolveModelDir(modelId: modelId)
        let fm = FileManager.default
        try? fm.removeItem(at: dir)
        states[modelId] = .idle
        return !fm.fileExists(atPath: dir.path)
    }

    // MARK: - Private

    private func resolveModelDir(modelId: String) -> URL {
        let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let root = support.appendingPathComponent("models")
        try? FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        return root.appendingPathComponent(modelId)
    }
}
